e2e/testcases/helm_sync_test.go

```diff
@@ -187,6 +187,76 @@ func TestHelmDefaultNamespace(t *testing.T) {
+//
+//	gcloud iam service-accounts add-iam-policy-binding --project=${GCP_PROJECT} \
+//	   --role roles/iam.workloadIdentityUser \
+//	   --member "serviceAccount:${GCP_PROJECT}.svc.id.goog[config-management-system/root-reconciler]" \
+//	   e2e-test-ar-reader@${GCP_PROJECT}.iam.gserviceaccount.com
+//
+// 4. The following environment variables are set: GCP_PROJECT, GCP_CLUSTER, GCP_REGION|GCP_ZONE.
+func TestHelmLatestVersion(t *testing.T) {
+	nt := nomostest.New(t,
+		nomostesting.WorkloadIdentity,
+		ntopts.Unstructured,
+		ntopts.RequireGKE(t),
+	)
+
+	rs := fake.RootSyncObjectV1Beta1(configsync.RootSyncName)
+	remoteHelmChart, err := helm.PushHelmChart(nt, privateSimpleHelmChart, privateSimpleHelmChartVersion)
+	if err != nil {
+		nt.T.Fatalf("failed to push helm chart: %v", err)
+	}
+
+	nt.T.Log("Update RootSync to sync from a private Artifact Registry")
+	nt.MustMergePatch(rs, fmt.Sprintf(`{"spec": {"sourceType": "%s", "helm": {"chart": "%s", "repo": "%s", "version": "", "auth": "gcpserviceaccount", "gcpServiceAccountEmail": "%s", "deployNamespace": "simple"}, "git": null}}`,
+		v1beta1.HelmSource, remoteHelmChart.ChartName, helm.PrivateARHelmRegistry, gsaARReaderEmail))
+	if err = nt.Watcher.WatchObject(kinds.Deployment(), "deploy-default", "simple",
+		[]testpredicates.Predicate{testpredicates.HasLabel("version", privateSimpleHelmChartVersion)}); err != nil {
+		nt.T.Error(err)
+	}
+
+	// helm-sync automatically detects and updates to the new helm chart version
+	newVersion := "2.5.9"
+	if err := remoteHelmChart.UpdateVersion(newVersion); err != nil {
+		nt.T.Fatal(err)
+	}
+	if err := remoteHelmChart.Push(); err != nil {
+		nt.T.Fatal("failed to push helm chart update: %v", err)
+	}
+	if err = nt.Watcher.WatchObject(kinds.Deployment(), "deploy-default", "simple",
+		[]testpredicates.Predicate{testpredicates.HasLabel("version", newVersion)}); err != nil {
+		nt.T.Error(err)
+	}
+}
+
+// TestHelmVersionRange verifies the Config Sync behavior for helm charts when helm.spec.version is specified as a range.
+// Helm-sync should pull the latest helm chart version within the range.
+func TestHelmVersionRange(t *testing.T) {
+	nt := nomostest.New(t, nomostesting.SyncSource, ntopts.Unstructured)
+
+	nt.T.Log("Create RootSync to sync from a public Helm Chart with specified version range")
+	rootSyncFilePath := "../testdata/root-sync-helm-chart-version-range-cr.yaml"
+	nt.T.Logf("Apply the RootSync object defined in %s", rootSyncFilePath)
+	nt.MustKubectl("apply", "-f", rootSyncFilePath)
+	err := nt.WatchForAllSyncs(nomostest.WithRootSha1Func(helmChartVersion("wordpress:15.4.1")),
+		nomostest.WithSyncDirectoryMap(map[types.NamespacedName]string{nomostest.DefaultRootRepoNamespacedName: "wordpress"}))
+	if err != nil {
+		nt.T.Fatal(err)
+	}
+
+	if err := nt.Validate("my-wordpress", "wordpress", &appsv1.Deployment{}); err != nil {
```

Comment:
Does this test not need to check a version label on the deployment?

Reply:
This is a public helm chart made by bitnami, so no, the deployment template does not have a version label.

The other test is a chart one we wrote ourselves, and I added a version label to it myself in this PR for the automatic update test, so that I could check that the new version did indeed get resynced to the cluster. 

This test however is just testing that we grab the latest version from a version range, so checking that the correct version was pulled down is sufficient here, which we do by checking the commit in L249.

Reply:
Ack, thank you for clarifying